File: src/determinism_check/determinism_check.erl
-module(determinism_check).
-behaviour(temporal_sdk_api_workflow_check).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../../docs/determinism_check.md"}.

-export([
    replay/3
]).
-export([
    is_deterministic/4
]).

-spec replay(CheckMod :: module(), Input :: binary(), ReplayInput :: binary()) ->
    temporal_sdk:replay_json_ret().
replay(CheckMod, Input, ReplayInput) ->
    % eqwalizer:ignore
    temporal_sdk:replay_task(
        cluster_1,
        "default",
        determinism_check_workflow,
        determinism_check_workflow,
        [
            {start_workflow_opts, [{input, [[Input]]}]},
            {replay_workflow_opts, [
                {worker_opts, [
                    {disable_telemetry, true},
                    {task_settings, [{deterministic_check_mod, CheckMod}]}
                ]},
                {task_overwrites, [{input, [[ReplayInput]]}]}
            ]}
        ]
    ).

%% -------------------------------------------------------------------------------------------------
%% behaviour implementation

-doc false.
is_deterministic(
    {ActualAwaitable, #{event_id := EventId, state := ActualState}},
    {ReceivedAwaitable, #{event_id := EventId, state := ReceivedState}},
    _ActualCommand,
    _ReceivedHistoryEvent
) ->
    test_state(ActualState, ReceivedState) andalso
        test_awaitable(ActualAwaitable, ReceivedAwaitable);
is_deterministic(_ActualAwaitable, _ReceivedAwaitable, _ActualCommand, _ReceivedHistoryEvent) ->
    false.

test_state(cmd, _State) -> true;
test_state(started, canceled) -> true;
test_state(State, State) -> true;
test_state(_ActualState, _ReceivedState) -> false.

test_awaitable({A}, {A}) -> true;
test_awaitable({A, _}, {A, _}) -> true;
test_awaitable({A, AT, _}, {A, AT, _}) -> true;
test_awaitable(_ActualAwaitable, _ReceivedAwaitable) -> false.

File: lib/determinism_check/determinism_check.ex
defmodule DeterminismCheck do
  @behaviour :temporal_sdk_api_workflow_check
  @external_resource "docs/determinism_check.md"
  @moduledoc TemporalSdk.Utils.exdoc!("docs/determinism_check.md")

  @spec replay(check_mod :: module(), input :: binary(), replay_input :: binary()) ::
          :temporal_sdk.replay_json_ret()
  def replay(check_mod, input, replay_input) do
    TemporalSdk.replay_task(
      :cluster_1,
      "default",
      DeterminismCheck.Workflow,
      DeterminismCheck.Workflow,
      start_workflow_opts: [input: [[input]]],
      replay_workflow_opts: [
        worker_opts: [
          disable_telemetry: true,
          task_settings: [deterministic_check_mod: check_mod]
        ],
        task_overwrites: [input: [[replay_input]]]
      ]
    )
  end

  # -------------------------------------------------------------------------------------------------
  # behaviour implementation

  @doc false
  def is_deterministic(
        {actual_awaitable, %{event_id: event_id, state: actual_state}},
        {replayed_awaitable, %{event_id: event_id, state: replayed_state}},
        _actual_command,
        _replayed_history_event
      ),
      do:
        test_state(actual_state, replayed_state) and
          test_awaitable(actual_awaitable, replayed_awaitable)

  def is_deterministic(
        _actual_awaitable,
        _replayed_awaitable,
        _actual_command,
        _replayed_history_event
      ),
      do: false

  defp test_state(:cmd, _state), do: true
  defp test_state(:started, :canceled), do: true
  defp test_state(state, state), do: true
  defp test_state(_actual_state, _replayed_state), do: false

  defp test_awaitable({a}, {a}), do: true
  defp test_awaitable({a, _}, {a, _}), do: true
  defp test_awaitable({a, at, _}, {a, at, _}), do: true
  defp test_awaitable(_actual_awaitable, _replayed_awaitable), do: false
end
